packages/expo-in-app-purchases/android/src/main/java/expo/modules/inapppurchases/BillingManager.java

```diff
@@ -253,27 +259,39 @@ public int getBillingClientResponseCode() {
   }
 
   /**
-   * Query purchases across various use cases and deliver the result in a formalized way through
-   * a listener
+   * Query both in app purchases and subscriptions and deliver the result in a formalized way 
+   * through a listener
    */
   public void queryPurchases(final Promise promise) {
     Runnable queryToExecute = new Runnable() {
+      BillingResult inAppBillingResult;
+      List<Purchase> purchases = new ArrayList<>();
+
       @Override
       public void run() {
-        PurchasesResult purchasesResult = mBillingClient.queryPurchases(SkuType.INAPP);
-
-        // If there are subscriptions supported, we add subscription rows as well
-        if (purchasesResult != null && areSubscriptionsSupported()) {
-          PurchasesResult subscriptionResult
-            = mBillingClient.queryPurchases(SkuType.SUBS);
+        mBillingClient.queryPurchasesAsync(SkuType.INAPP, new PurchasesResponseListener() {
+          @Override
+          public void onQueryPurchasesResponse(@NonNull BillingResult billingResult, @NonNull List<Purchase> inAppPurchases) {
+            inAppBillingResult = billingResult;
+            if (billingResult.getResponseCode() == BillingResponseCode.OK) {
+              purchases.addAll(inAppPurchases);
+            }
 
-          if (subscriptionResult != null && subscriptionResult.getResponseCode() == BillingResponseCode.OK) {
-            purchasesResult.getPurchasesList().addAll(
-              subscriptionResult.getPurchasesList());
+            if (areSubscriptionsSupported()) {
+              mBillingClient.queryPurchasesAsync(SkuType.SUBS, new PurchasesResponseListener() {
```

Comment:
Instead of nesting the async calls it would be better to run the two queryPurchasesAsync calls in parallel.

Reply:
Yeah had that originally, but we need to keep track of which listener finishes last, to call `onQueryPurchasesFinished`. Anyways, the extra lines of code are probably worth parallelizing